### ISSUE TITLE
Difference between FromField and QueryRunnerColumnDefault

### ISSUE THREAD

**Issue by saurabhnanda** [tomjaguarpaw/haskell-opaleye#246]
What is a scenario where QueryRunnerColumnDefault will be defined as something **other than:**

```
instance QueryRunnerColumnDefault PGText Gender where
  queryRunnerColumnDefault = fieldQueryRunnerColumn
```

Conceptually, the DB=>Haskell mapping comes from `FromField`. What is `QueryRunnerColumnDefault` really used for?

**Comment by tomjaguarpaw** [comment]
Well, there are two in `RunQuery.hs`!

https://github.com/tomjaguarpaw/haskell-opaleye/blob/master/src/Opaleye/Internal/RunQuery.hs#L201

**Comment by tomjaguarpaw** [comment]
`QueryRunnerColumnDefault` is a somewhat "augmented" version of `FromField`.  `FromField` *only* tells you how to convert something in the Postgres wire protocol into a Haskell value.  `QueryRunnerColumnDefault` *also* tells you which `Column` types can be converted into that Haskell type.

Thus if there are two different instances with the same Haskell type, e.g.

* `instance QueryRunnerColumnDefault T.PGText String`, and

* `instance QueryRunnerColumnDefault T.PGJson String`

only one of them can come from a `FromField` instance (the former, in this example).

**Comment by saurabhnanda** [comment]
Thanks for trying to explain, but it still isn't clear to me. Are you trying to point out the need for two functions;

* ByteString -> Haskell
* Column x -> Haskell

If yes, then the question changes to why do we need two functions in the first place? And why isn't it  ByteString -> Column x  and then column x -> Haskell?

**Comment by saurabhnanda** [comment]
Also, what's with the cryptic names? eg. `QueryRunnerColumnDefault` -- what exactly is "Default" here? What is "QueryRunner"?

**Comment by tomjaguarpaw** [comment]
The name "QueryRunnerColumnDefault" is made up of four pieces:

* "QueryRunner" because it allows you to run queries
* "Column" because it represents how a single `Column` is decoded
* "Default" because it is the default way of decoding that column type.  You can always supply others if you like.

**Comment by saurabhnanda** [comment]
I'm unable to complete [this section of the tutorial](http://haskell-webapps.readthedocs.io/en/latest/docs/opaleye/advanced-db-mapping.html#core-mechanism-for-mapping-custom-haskell-types-to-pg-types) because I myself dont understand the need for `QueryRunnerColumnDefault` over & above `FromField`. Can you please help me understand this better?

**Comment by sras** [comment]
@saurabhnanda I think that the need for QueryRunnerColumnDefault is that the FromField typeclass does not contain the type of the Database column in its definition

```
class FromField a where
  fromField :: FieldParser a
```

As you can see, it does not say which type of columns can be converted to `a`. I think postgresql-simple does not provide a notion of database column types in the type system itself (though you can query the field type using the first parameter to fromField function).

So in short, I think QueryRunnerColumnDefault is required because Opaleye wants to enable type system checks to ensure that you are not reading from an incompatible database column.

Even in this case, you have got to make sure that you are using a compatible fromField implementation (if you are using it to implement QueryRunnerColumnDefault), or else your code will error out at runtime.

**Comment by tomjaguarpaw** [comment]
Sandeep is correct.  Can I be of any more assistance with this question?

**Comment by saurabhnanda** [comment]
@tomjaguarpaw  thanks for following up. Does this mean that pg-simple will allow a pg bool to be parsed as a Haskell Int, and let it result in a potential conversion failure at runtime?

Is it possible to have a single type-class managed by Opaleye which takes care of fromfield's machinery as well?

**Comment by tomjaguarpaw** [comment]
> Does this mean that pg-simple will allow a pg bool to be parsed as a Haskell Int, and let it result in a potential conversion failure at runtime?

Absolutely!

> Is it possible to have a single type-class managed by Opaleye which takes care of fromfield's machinery as well?

I'm not sure what you mean.  Do you mean "could `FromField` be entirely replaced by `QueryRunnerColumnDefault`"?

**Comment by saurabhnanda** [comment]
> I'm not sure what you mean. Do you mean "could FromField be entirely replaced by QueryRunnerColumnDefault"?

Yes. Although I would canvass for s better name :)

**Comment by tomjaguarpaw** [comment]
Yes, arguably `postgresql-simple` would be a better library if it represented the Postgres type of columns it is decoding.  The counter argument would be that it is extra complexity and only represents information that `postgresql-simple` users are comfortable doing without.

> better name

I'm very happy to take suggestions for better names.

**Comment by saurabhnanda** [comment]
Still following up on this. Somehow, my gut feels, this is more complicated than it needs to be. 

Reading the documentation at https://github.com/tomjaguarpaw/haskell-opaleye/blob/master/src/Opaleye/Internal/RunQuery.hs#L126 makes me wonder if it really is not possible to shield programmers from the quirks of pg-simple and give one simple way to read/write custom types via Opaleye. 

Also the docs say the following -- 

> the default way to turn a @pgType@ result from the database into a Haskell value of type @haskellType@.

What is an example of a **non-default** way of pg=>Haskell conversion? How does one use it?

**Comment by tomjaguarpaw** [comment]
> Reading the documentation at https://github.com/tomjaguarpaw/haskell-opaleye/blob/master/src/Opaleye/Internal/RunQuery.hs#L126 makes me wonder if it really is not possible to shield programmers from the quirks of pg-simple and give one simple way to read/write custom types via Opaleye.

Yes, I would prefer one simple approach.  I haven't really spent much time on those docs.

> What is an example of a non-default way of pg=>Haskell conversion? How does one use it?

A non-default way would involve constructing a `QueryRunner columns haskells` yourself and then using the explicit version of `runQuery`:

    runQueryExplicit :: QueryRunner columns haskells -> Connection -> Query columns -> IO [haskells]

**Comment by saurabhnanda** [comment]
> Yes, I would prefer one simple approach. I haven't really spent much time on those docs.

I was referring to the core mechanism, not just documenting jy better. Is it possible to have a single type-class to be implemented to teach Opaleye how to deal with custom data-types?

> A non-default way would involve constructing a QueryRunner columns haskells yourself and then using the explicit version of runQuery:

Would it be correct to say that it's not really about the *columns default* buy the *default query runner*? If I want a PG type to map to multiple possible Haskell types, then each of them necessarily requires a new query runner to be implemented. If this understanding is correct, then my point about confusing names is more pertinent. 

Now, I have finally understood the current mechanism, isn't the ENUM mapping given at https://github.com/tomjaguarpaw/haskell-opaleye/issues/197#issuecomment-245175004 partially incorrect?  It is parsing a new PG type, say `gender_type` to PGText, thus causing loss in type fidelity. This will allow me to write SQL queries that compare a gender columns to any text column. Is this correct?

**Comment by tomjaguarpaw** [comment]
> I was referring to the core mechanism, not just documenting jy better. Is it possible to have a single type-class to be implemented to teach Opaleye how to deal with custom data-types?

`QueryRunnerColumnDefault` is the single typeclass that teaches Opaleye how to deal with custom datatypes.  It turns out that for technical typeclass reasons (and I can't remember the reason right now) I needed both `QueryRunnerColumnDefault` and `QueryRunner` to exist.

> Now, I have finally understood the current mechanism, isn't the ENUM mapping given at #197 (comment) partially incorrect? It is parsing a new PG type, say gender_type to PGText, thus causing loss in type fidelity. This will allow me to write SQL queries that compare a gender columns to any text column. Is this correct?

Your analysis is correct.  It would probably be better to define

````
data TenantStatus = Active | Inactive
data PGTenantStatus

type User = User' TenantStatus
type RUserColumn = User' (Column PGTenantStatus)
type WUserColumn = User' (Column PGTenantStatus)
````

**Comment by saurabhnanda** [comment]
> QueryRunnerColumnDefault is the single typeclass that teaches Opaleye how to deal with custom datatypes. It turns out that for technical typeclass reasons (and I can't remember the reason right now) I needed both QueryRunnerColumnDefault and QueryRunner to exist.

Does this mean we don't really need to implement a `FromField` instance as described at https://github.com/tomjaguarpaw/haskell-opaleye/issues/197#issuecomment-245175004 ? I'm confused now!

Request you to please explain how this is working. Is some boilerplate code missing?

```
data TenantStatus = Active | Inactive
data PGTenantStatus

type User = User' TenantStatus
type RUserColumn = User' (Column PGTenantStatus)
type WUserColumn = User' (Column PGTenantStatus)
```

**Comment by tomjaguarpaw** [comment]
You still need the `FromField` instance.  It should be the same as before.

**Comment by saurabhnanda** [comment]
> You still need the FromField instance. It should be the same as before.

Which brings me *back* to the point I was making earlier: 

> Is it possible to have a single type-class to be implemented to teach Opaleye how to deal with custom data-types?

From a programmer's standpoint, is there any benefit in exposing extra complexity about how Opaleye deals with pg-simple and why `FromField` is not part of Opaleye, etc. etc?

**Comment by saurabhnanda** [comment]
Also, any thoughts on:

> Would it be correct to say that it's not really about the column default but the default query runner? If I want a PG type to map to multiple possible Haskell types, then each of them necessarily requires a new query runner to be implemented.

**Comment by tomjaguarpaw** [comment]
> From a programmer's standpoint, is there any benefit in exposing extra complexity about how Opaleye deals with pg-simple and why FromField is not part of Opaleye, etc. etc?

Oh I see. Well, maybe not.  But I just didn't do it that way.

**Comment by tomjaguarpaw** [comment]
>    Would it be correct to say that it's not really about the column default but the default query runner? If I want a PG type to map to multiple possible Haskell types, then each of them necessarily requires a new query runner to be implemented.

**Comment by tomjaguarpaw** [comment]
>    Would it be correct to say that it's not really about the column default but the default query runner? If I want a PG type to map to multiple possible Haskell types, then each of them necessarily requires a new query runner to be implemented.

Perhaps there's bad naming here.  It's not the "column default" is the "default query runner for individual columns".

**Comment by tomjaguarpaw** [comment]
(I added a better `FromField` instance:  https://github.com/tomjaguarpaw/haskell-opaleye/issues/197#issuecomment-264797033

Hopefully you can see how to transfer this to your tenant example.)

**Comment by saurabhnanda** [comment]
> (I added a better FromField instance: #197 (comment)
> Hopefully you can see how to transfer this to your tenant example.)

Thanks. Follow-up comment regarding this topic on that issue now.

**Comment by tomjaguarpaw** [comment]
If you need more help with this please reopen.
